SQL-804: Make MongoSQLResultSet.getObject().toString() consistent with getString()

mattChiaravalloti opened the pull request:
This PR updates `MongoSQLResultSet.getObject()` to return a value such that calling `toString()` on it is consistent with calling `MongoSQLResultSet.getString()`. Both versions now produce what `getString()` was already returning. As discussed in standup, this was accomplished by wrapping the result of `getObject()` for `Types.OTHER` types in a new wrapper class that overrides `toString()` as needed.

~Note: we separately discussed having all `Types.OTHER` types stringify to extended JSON, but that change is outside the scope of this ticket. We technically haven't even decided if that is the direction we want to go as that is still up for debate in the design doc. Therefore, this ticket does not introduce any new behavior to the stringification -- it exactly matches what `getString()` was already doing.~

After the design discussion and seeing this ticket is included in the epic, I will update stringification to use extended json for the OTHER types as part of this ticket.

mattChiaravalloti commented on src/main/java/com/mongodb/jdbc/ExtJsonValue.java:
```diff
@@ -0,0 +1,109 @@
+            case SYMBOL:
+            case TIMESTAMP:
+            default:
+                return "<invalid>"; // TODO - what should we do for these types?
```
This is the biggest question: what should we do with these other types? In `getString()` these cases throw exceptions. We don't have that option here, so I'm not sure what we want to do. This string is just a placeholder until we decide. We could default to `this.v.toString()` for these for now, pending what we decide as part of the design process.

mattChiaravalloti replied:
resolved offline: going to change these to use their extended json representation.

mattChiaravalloti commented on src/test/java/com/mongodb/jdbc/MongoSQLMock.java:
```diff
@@ -35,10 +35,12 @@
     // foo.d
     protected static int ANY_COL = 7;
+    // foo.doc
+    protected static int DOC_COL = 8;
```
the labels have to be alphabetically ordered, which is why this is the new 8 as opposed to being introduced as 10

mattChiaravalloti commented on src/test/java/com/mongodb/jdbc/MongoSQLResultSetTest.java:
```diff
@@ -262,17 +277,40 @@ void testGetStringValues() throws Exception {
+        assertEquals("[5, 6, 7]", mongoSQLResultSet.getString(ARRAY_COL_LABEL));
+        assertEquals("{\"c\": 5}", mongoSQLResultSet.getString(DOC_COL_LABEL));
+
+        // Check that getObject().toString() matches getString()
```
This is the primary place where I added new testing. By adding the document value `doc`, I am covering the two cases this currently primarily impacts: arrays and documents.

nbagnard replied:
We should aim for covering all BSON types and on top of checking that getString and getObjectToString match, we should also make sure that the values are valid relaxed extended JSON. This can be done with something like `Document.parse("{\"val\":" + extendedJsonVal + "}")`. If the parsing does not fail, then it means extendedJsonVal is a valid extended json value.

nbagnard commented on src/main/java/com/mongodb/jdbc/ExtJsonValue.java:
```diff
@@ -0,0 +1,109 @@
+            case NULL:
+                return null;
+            case OBJECT_ID:
+                return this.v.asObjectId().getValue().toString();
```
I did a run a quick test and BsonObjectId.getValue().toString() does not return the relaxed extended Json representation.
BsonObjectId.getValue().toString() => 62478aa9cc79dc251e626f7d
Relaxed extended JSON representation (Same a canonical for oid) => {"$oid": "62478aa9cc79dc251e626f7d"}

Do we want a special treatment for OIDs? IMO the extended JSON representation would be good there because it makes it clear its not a simple string. I thought this was something we wanted to aim for, but I remember OIDs are a bit of a muddy ground so maybe I missed something.

mattChiaravalloti replied:
This was initially implemented before we decided this ticket would cover all types (not just object and array) so it does not have the work we discussed in standup on friday! will work on that today. @bucaojit I like that idea, but I will file a follow-up ticket to introduce that.

mattChiaravalloti replied:
done:
1. PR updated to have extended json stringification for all types.OTHER types, and
2. filed [SQL-813](https://jira.mongodb.org/browse/SQL-813) to follow-up on Oliver's suggestion

bucaojit commented on src/test/java/com/mongodb/jdbc/MongoSQLMock.java:
```diff
@@ -266,6 +288,10 @@ BsonDocument generateRow() {
         foo.put(ARRAY_COL_LABEL, array);
 
+        BsonDocument fooSubDoc = new BsonDocument();
+        fooSubDoc.put(INT_COL_LABEL, new BsonInt32(5));
```
Would it be possible to add a BsonValue that has type information such as objectId, that would help confirm that the toString() is writing the extended json value.

mattChiaravalloti replied:
now that we decided this ticket will cover stringifying all types as extended json, going to update the unit tests to cover all types

mattChiaravalloti replied:
done

mattChiaravalloti replied:
@nbagnard I'm not sure I understand what your suggestion is here. What specific test methods are you suggesting I update/remove/addd? Can you provide an example of what you're envisioning?

mattChiaravalloti replied:
done. the code turned out slightly longer in each case: 96 lines (new) vs. 95 (old) for schema, and 83 lines (new) vs 36 (old) for data, but we got to cut the comment out from the schema method and I think it should be more readable now. lmk what you think!

mattChiaravalloti replied:
Yea, maybe part of the JDBC Test Infrastructure epic could be cleaning up some of the clutter we introduced to the unit tests over the last few months 👍

bucaojit reviewed (changes requested):
Great work adding the ExtJsonValue class and making the toString() outputs consistent.

rychipman commented on src/main/java/com/mongodb/jdbc/ExtJsonValue.java:
```diff
@@ -0,0 +1,109 @@
+ * <p>The driver's BsonValue class is abstract and intentionally cannot be extended by third
+ * parties. The driver explains this is to keep the BSON type system closed.
+ */
+public class ExtJsonValue {
```
One drive-by request from me: could we make this something more generically named than `ExtJsonValue` (perhaps something like `MongoSQLValue`)? If we ever want to do something in the future like serialize these strings differently or make string serialization configurable, this class name might become outdated (but we won't be able to change it without it _technically_ being backwards breaking)

mattChiaravalloti replied:
sure thing. I wasn't really sure what name would be good here. I'll make this change today.

nbagnard replied:
The proposed name on the design doc was "MongoSQLBsonValue" but "MongSQLValue" is just as good.

mattChiaravalloti replied:
done, but could still change this to `MongoSQLBsonValue`. Sorry I missed that in the design 😓

mattChiaravalloti replied:
Updated to be consistent with design doc.

mattChiaravalloti commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/IntegrationTestUtils.java:
```diff
@@ -673,18 +675,23 @@ public static boolean compareRow(List<Object> expectedRow, ResultSet actualRow)
                         }
+                    } else {
+                        throw new IllegalArgumentException(
+                                "unsupported expected value class: " + expected_obj.getClass());
```
@bucaojit I left the previous case as `else if (expected_obj instanceof BsonValue)` so we can throw this more meaningful exception in case someone someday adds a test that has some other class as an expected result.

mattChiaravalloti commented on src/main/java/com/mongodb/jdbc/MongoSQLValue.java:
```diff
@@ -0,0 +1,108 @@
+                        ENCODER_CONTEXT);
+                w.flush();
+                return w.toString();
+            case END_OF_DOCUMENT:
+            default:
+                return "<invalid>";
```
Still not sure what to do for `END_OF_DOCUMENT` (the `default` is technically unreachable since the switch statement exhaustively matches all variants of the `BsonType` enum. Should I return a static string like I have here, or perhaps `this.v.toString()`, or should we throw an `IllegalStateException` or other runtime exception?

mattChiaravalloti replied:
Updated to return `v.toString()`

mattChiaravalloti commented on src/test/java/com/mongodb/jdbc/MongoSQLMock.java:
```diff
@@ -284,4 +361,239 @@ MongoJsonSchemaResult generateSchema() {
         return schemaResult;
     }
+
+    protected static MongoJsonSchema generateMongoJsonSchemaAllTypes() {
```
I added a "collection" to the mock that contains all bson types. I added this `generateMongoJsonSchemaAllTypes` method and the `generateRowAllTypes` method so we could create a mock result set that has all types. I only use this "collection" for testing the `getString()` and `getObject().toString()` behavior since that is all that is really in scope for this ticket.

mattChiaravalloti commented on src/test/java/com/mongodb/jdbc/MongoSQLResultSetMetaDataTest.java:
```diff
@@ -37,17 +37,17 @@ void initMocks() {
     @Test
     void testGetColumnCount() throws SQLException {
-        assertEquals(9, MongoSQLResultSetMetaDataTest.resultSetMetaData.getColumnCount());
+        assertEquals(10, MongoSQLResultSetMetaDataTest.resultSetMetaData.getColumnCount());
```
Now that I added the "all types collection", should I rollback these initial test updates that simply added a document to the previous mock collection? Technically this change is now outside the scope of this ticket and doesn't add any new coverage for work done _in this ticket_, but does add some additional coverage we didn't previously have in the unit tests. Since it is already written, I'd gladly leave it. But I can also take it away very easily.

mattChiaravalloti replied:
Yea it looks like the initial tests for result set included only types that didn't map to types.OTHER, but that's just speculation.

mattChiaravalloti commented on src/test/java/com/mongodb/jdbc/MongoSQLResultSetTest.java:
```diff
@@ -285,6 +320,186 @@ void testGetStringValues() throws Exception {
+                mongoSQLResultSetAllTypes.getObject(ALL_DOC_COL_LABEL).toString());
+        assertEquals(
+                "[7, 8, 9]", mongoSQLResultSetAllTypes.getObject(ALL_ARRAY_COL_LABEL).toString());
+        // TODO: Java driver doesn't seem to serialize binary using extended json
+        //        assertEquals("{\"$binary\": {\"base64\": \"\", \"subType\": \"0\"}}", mongoSQLResultSetAllTypes.getObject(ALL_BINARY_COL_LABEL).toString());
```
Still need to look into why this is happening. Even using the java driver's extended json serialization methods results in `BsonBinary` being stringified in the wrong format:
```
Expected : {"$binary": {"base64": "", "subType": "0"}}
Actual   : BsonBinary{type=0, data=[]}
```
Could be a misunderstanding of the java driver api, or could be a bug in the java driver. My current guess is it is the latter case, since the binary codepath is the same as all the types.OTHER bson types' codepath and they all get stringified as ext json.

mattChiaravalloti replied:
Oh I see. Obvious mistake. The `getObject()` method returns a `BsonBinary` for binary types because bson binary maps to `Types.BINARY`, not `Types.OTHER`. Easy fix.

nbagnard commented on src/test/java/com/mongodb/jdbc/MongoSQLResultSetTest.java:
```diff
@@ -285,6 +320,186 @@ void testGetStringValues() throws Exception {
+        assertNull(new MongoSQLValue(new BsonNull()).toString());
+
+        // Test getObject().toString() is valid relaxed extended JSON
+        // TODO
```
I don't think that we are testing that the value are valid extended JSON value which are strictly matching the extended JSON spec. We are only testing that both getString and getObject.toString() are the same. If we were testing that the values are valid extended JSON, then you would see a failure for the Timestamp.
Try this out :
long unixTime = System.currentTimeMillis();
BsonDateTime timestamp = new BsonDateTime(unixTime);
Date date = new Date(timestamp.getValue());
extendedJsonVal = printExtendedJsonValue(timestamp, JsonMode.RELAXED);
System.out.println(dateFormat.format(date));
System.out.println(extendedJsonVal);

You'll notice that the 2 string representations are different.

Also, the date from timestamp.getValue() is also not in UTC even though the timezone implies it is, while the extended Json value is in UTC. I'll file a separate ticket for this.

mattChiaravalloti replied:
Yea sorry, noted on another thread: I didn't know we wanted to change getString() for BSON types that correspond to non-Types.OTHER JDBC types. However, even after we update to make sure getString()/getObject().toString() return extended json representations for all types, I'm not sure what additional coverage this additional test will provide.

We'll have the following assertions for every bson type:
```
assertEquals("{$oid: ...}", rs.getString(OID_COL));
assertEquals("1.2", rs.getString(DOUBLE_COL));
assertEquals("{$date: ...}, rs.getString(DATE_COL));
... // etc

assertEquals("$oid: ...}", rs.getObject(OID_COL).toString());
assertEquals("1.2", rs.getObject(DOUBLE_COL).toString());
assertEquals("{$date: ...}, rs.getObject(DATE_COL).toString());
// etc
```
Two assertions for each bson type where the expected string is the extended json representation. Since the expected string is already the extended json string representation, isn't that sufficient to show that the output of getString()/getObject().toString() is _valid_ extended json? It seems the additional test to ensure the output is valid would really just be testing that our expected strings in the other tests are valid.

nbagnard replied:
It would if we could be 100% sure that the string we put there is the correct extended json representation and using an extended Json compatible encoder works to reconstruct the BsonValue. As you can see, it is not the case for date_time and Decimal128 right now. Let's see after we discuss if we are indeed using the extended representation for every data type or if the others in the team also thought this was just for data types mapping to Other. After that, we can see how to make maybe change this test slightly to cover both it if needed.

nbagnard commented on src/main/java/com/mongodb/jdbc/MongoSQLValue.java:
```diff
@@ -0,0 +1,108 @@
+            case DATE_TIME:
+                Date d = new Date(this.v.asDateTime().getValue());
+                return dateFormat.format(d);
+            case DECIMAL128:
```
Same as for Date. This will not return the relaxed extended JSON representation.
The value we provide is : _121684.45864468_
The relaxed extended JSON representation is : _{"$numberDecimal": "121684.45864468"}_

mattChiaravalloti replied:
done

mattChiaravalloti commented on src/main/java/com/mongodb/jdbc/MongoSQLValue.java:
```diff
@@ -0,0 +1,108 @@
+        switch (this.v.getBsonType()) {
+            case BOOLEAN:
+                return this.v.asBoolean().getValue() ? "true" : "false";
+            case DATE_TIME:
```
Oh, I just misunderstood. I was under the impression we just wanted the extended json string representation for types that map to Types.OTHER, not for all types. I can make the update to do extended json for all types.

mattChiaravalloti commented on src/main/java/com/mongodb/jdbc/MongoSQLResultSet.java:
```diff
@@ -388,36 +339,12 @@ protected String getString(BsonValue o) throws SQLException {
-                        ENCODER_CONTEXT);
-                writer.flush();
-                return writer.toString();
             case BINARY:
```
done (updated `getString` to support all types)

mattChiaravalloti commented on src/main/java/com/mongodb/jdbc/MongoSQLResultSet.java:
```diff
@@ -138,7 +119,7 @@ protected Object getObject(BsonValue o, int columnType) throws SQLException {
                 return getInt(o);
             case Types.BINARY:
-                return o.asBinary();
+                return new MongoSQLValue(o.asBinary());
```
It does feel a bit strange to only wrap _some_ types in `getObject` with `MongoSQLValue`. It may not be totally clear to future maintainers why some are wrapped (Binary, Decimal, Other, Timestamp) and some are not (all others that do not throw an exception). The distinction right now is that anything that needs special extended json stringification is wrapped. Not sure if that's the best criteria. Maybe we should wrap everything?

nbagnard commented on src/test/java/com/mongodb/jdbc/MongoSQLResultSetTest.java:
```diff
@@ -285,6 +320,192 @@ void testGetStringValues() throws Exception {
     }
 
+    @Test
+    public void testGetStringAllTypes() throws Exception {
```
I looked into this a bit more and I have a suggestion which I hope can make both of us happy.
I agree that adding an extra test to verify that the String is a valid extended JSON value seems a bit too much.
What we could do is : 
- Extract the values out in separate variables
- Use these snippets in generateRowAllTypes
- Use the them snippet as the expected value

This way, we know that the values are valid extended JSON because they were parsed without any issue by generateRowAllTypes , we verify it's the correct expected value for getString and because testGetObjectToStringMatchesGetString checks that getString == getObject.toString we are good all the way.

Below is a partial example :
`    
public abstract class MongoSQLMock {
...
protected static final String ALL_DOUBLE_COL_VAL = "5";
    protected static final String ALL_TIMESTAMP_COL_VAL = "{\"$timestamp\": {\"t\": 1412180887, \"i\": 1}}";
 ....
    static BsonDocument generateRowAllTypes() {
        String doc =
                    "{"
                        + "    \"all\": {"
                         ....
                         + "        \"timestamp\": "
                        +  ALL_TIMESTAMP_COL_VAL
                        + ","
                        + "        \"long\": "
                        + ALL_DOUBLE_COL_VAL
                        + ","
                       ....
    }
...
}

--------------------------------------------------------------------

class MongoSQLResultSetTest extends MongoSQLMock {
....
    @Test
    public void testGetStringAllTypes() throws Exception {
    ....
            assertEquals( 
                ALL_TIMESTAMP_COL_VAL,
                mongoSQLResultSetAllTypes.getString(ALL_TIMESTAMP_COL_LABEL));
            assertEquals(
               ALL_DOUBLE_COL_VAL, 
               mongoSQLResultSetAllTypes.getString(ALL_LONG_COL_LABEL));
    ....
    }
...
}
`

mattChiaravalloti replied:
done.

nbagnard commented on src/test/java/com/mongodb/jdbc/MongoSQLMock.java:
```diff
@@ -284,4 +358,191 @@ MongoJsonSchemaResult generateSchema() {
+
+    protected static MongoJsonSchema generateMongoJsonSchemaAllTypes() {
+        String schema =
+                ""
```
Nit: I think this empty string can be safely removed. Or does it come from Spotless?

nbagnard commented on src/test/java/com/mongodb/jdbc/MongoSQLMock.java:
```diff
@@ -284,4 +358,191 @@ MongoJsonSchemaResult generateSchema() {
+
+    static BsonDocument generateRowAllTypes() {
+        String doc =
+                ""
```
Nit: I think this empty string can be safely removed. Or does it come from Spotless?

nbagnard commented on src/main/java/com/mongodb/jdbc/MongoSQLValue.java:
```diff
@@ -0,0 +1,104 @@
+            JsonWriterSettings.builder().outputMode(JsonMode.RELAXED).build();
+    static final EncoderContext ENCODER_CONTEXT = EncoderContext.builder().build();
+
+    // dateFormat cannot be static due to a threading bug in the library.
```
I think this is probably unused now, isn't it?

nbagnard reviewed (changes requested):
Looking good!
I only have a few comments (clean-up stuff) and one suggestion for making sure the produced values are extended JSON :)

nbagnard commented:
(Putting in the right PR)
ah, bummer, I looked at the spec again because I was surprised that all integer related types are mapped to getInt and I think that there are a couple of types for which getObject().toString will not match getString if we want to match the JDBC spec.
See B.3 JDBC Types Mapped to Java Object Types from the spec:

    DECIMAL/NUMERIC => java.math.BigDecimal
    BINARY/VARBINARY/LONGVARBINARY => byte[]
    TIMESTAMP => java.sql.Timestamp

I would still keep the getString() value to be the extended JSON string though because I did not find anything regarding what getString is supposed to return. The spec only says that it should return a value for all mainstream SQL/JDBC types (we were not obligated to return a String for OTHER for example according to the spec, but it looks like they just missed it, same for Null).

It seems safer to follow the spec for getObject, just in case someone's application uses getObject unconditionally and then work directly on the objects with the assumption it match the spec and that getObject for a TIMESTAMP return a java.sql.timestamp for example.
WDYT?

mattChiaravalloti commented:
> (Putting in the right PR) ah, bummer, I looked at the spec again because I was surprised that all integer related types are mapped to getInt and I think that there are a couple of types for which getObject().toString will not match getString if we want to match the JDBC spec. See B.3 JDBC Types Mapped to Java Object Types from the spec:
> 
> ```
> DECIMAL/NUMERIC => java.math.BigDecimal
> BINARY/VARBINARY/LONGVARBINARY => byte[]
> TIMESTAMP => java.sql.Timestamp
> ```
> 
> I would still keep the getString() value to be the extended JSON string though because I did not find anything regarding what getString is supposed to return. The spec only says that it should return a value for all mainstream SQL/JDBC types (we were not obligated to return a String for OTHER for example according to the spec, but it looks like they just missed it, same for Null).
> 
> It seems safer to follow the spec for getObject, just in case someone's application uses getObject unconditionally and then work directly on the objects with the assumption it match the spec and that getObject for a TIMESTAMP return a java.sql.timestamp for example. WDYT?

discussed offline -- I updated the `getObject` method to return the instances of the appropriate classes for the ones you listed above (although `VARBINARY` AND `LONGVARBINARY` are still left as invalid for now. I believe there is another ticket to update the JDBC type for BSON binary).

This is now implemented and up to date!

bucaojit reviewed (approved):
LGTM!  
Incredible work, and great to see all the test improvements. 

There is one unit test failing for the date column: 
```
AssertionFailedError: expected: <2020-12-25 12:13:14.0> but was: <2020-12-25 17:13:14.0>
```

nbagnard commented on src/test/java/com/mongodb/jdbc/MongoSQLResultSetTest.java:
```diff
@@ -285,6 +320,175 @@ void testGetStringValues() throws Exception {
+        assertEquals(
+                ALL_MAX_KEY_COL_VAL, mongoSQLResultSetAllTypes.getString(ALL_MAX_KEY_COL_LABEL));
+
+        // Note that for "string" and "long", the ALL_X_COL_VAL must use extended json style-syntax in the
```
Using this code in the Mock will let use both canonical and relaxed extended Json represention :
Document d = Document.parse(_extendedJson_);
BsonDocument bsonDoc =  d.toBsonDocument(Document.class, MongoClient.getDefaultCodecRegistry());

For example, the snippet above accepts {"val": 1} and {"val": {"$numberInt": "1"}} as the _extendedJson_ value.

mattChiaravalloti replied:
discussed offline: I updated `MongoSQLBsonValue` to stringify all BsonValues using relaxed extended json. We will discuss in standup whether this is the behavior we want for bson strings, which will now be delimited with `"` characters.

nbagnard commented on src/main/java/com/mongodb/jdbc/MongoSQLResultSet.java:
```diff
@@ -132,136 +114,65 @@ protected Object getObject(BsonValue o, int columnType) throws SQLException {
-                break;
             case Types.BOOLEAN:
                 return getBoolean(o);
             case Types.CHAR:
```
Looks you missed moving CHAR with LONGVARCHAR, VARCHAR, etc....

mattChiaravalloti replied:
done

nbagnard commented on src/main/java/com/mongodb/jdbc/MongoSQLResultSet.java:
```diff
@@ -132,136 +114,65 @@ protected Object getObject(BsonValue o, int columnType) throws SQLException {
-                break;
-            case Types.TINYINT:
-                return getInt(o);
             case Types.VARBINARY:
```
Could we move VARBINARY and LONGVARBINARY with BINARY, the same way we group BIGINT, INT, SMALLINT, TINYINT and CHAR, VARCHAR, LONGVARCHAR?

mattChiaravalloti replied:
I can make this change, but does this really matter? The only JDBC types that this method will ever encounter are the ones associated with BSON types in `BsonTypeInfo`. `BsonTypeInfo` does not have any reference to `Types.VARBINARY` or `Types.LONGVARBINARY`.

(The same thing is true for `Types.SMALLINT`, `Types.TINYINT`, `Types.BOOLEAN`, `Types.FLOAT`, `Types.Numeric`, `Types.CHAR`, `Types.VARCHAR`, `Types.LONGNVARCHAR`, `Types.NCHAR`, `Types.NVARCHAR`, and `Types.Real`, actually).

nbagnard replied:
No, it does not matter technically until we change any mapping. But it makes sense to at least have the same behaviour for all types with either only providing a value for the SQL type we support or always try to get the value with the getter for the "group". Otherwise, I feel like it opens the door for questions about why treating String related types differently than binary related type? Even if the answer is "Why not" :P

mattChiaravalloti replied:
good enough haha, updated!

nbagnard replied:
thanks :)

nbagnard commented on src/main/java/com/mongodb/jdbc/MongoSQLResultSet.java:
```diff
@@ -132,136 +114,65 @@ protected Object getObject(BsonValue o, int columnType) throws SQLException {
-            case Types.ARRAY:
-                // not supported
-                break;
             case Types.BIGINT:
```
BIGINT should return a Long. Otherwise we might encounter overflow issues. Especially since we correctly map BSON_LONG to BIGINT.

mattChiaravalloti replied:
done

nbagnard commented on src/test/java/com/mongodb/jdbc/MongoSQLResultSetTest.java:
```diff
@@ -285,6 +320,175 @@ void testGetStringValues() throws Exception {
+                        .toString());
+        assertEquals(
+                ALL_INT_COL_VAL, mongoSQLResultSetAllTypes.getObject(ALL_INT_COL_LABEL).toString());
+        assertEquals(
```
This should not pass because mongoSQLResultSetAllTypes.getObject(ALL_TIMESTAMP_COL_LABEL) will return a java.sql.Timestamp and toString on it should return yyyy-mm-dd hh:mm:ss.fffffffff, not an extended JSON value like {"$timestamp": {"t": 1412180887, "i": 1}}.
I am a little surprised Evergreen is happy. Could you double-check? I am not sure what I missed here ?🧐

mattChiaravalloti replied:
Discussed offline: bson timestamp != bson date

nbagnard commented on src/test/java/com/mongodb/jdbc/MongoSQLResultSetTest.java:
```diff
@@ -285,6 +320,175 @@ void testGetStringValues() throws Exception {
+                mongoSQLResultSetAllTypes
+                        .getObject(ALL_JAVASCRIPT_WITH_SCOPE_COL_LABEL)
+                        .toString());
+        assertEquals(
```
Same here, I am not sure how this assert passes since getString should return {"$timestamp": {"t": 1412180887, "i": 1}} and getObject.toString should return something like "2020-12-25 12:13:14.0" (value formatted with yyyy-mm-dd hh:mm:ss.fffffffff)

nbagnard replied:
Got it, thanks Matt for the extra details on the fun of Bson.Timestamp :)

mattChiaravalloti replied:
discussed offline: bson timestamp != bson date

nbagnard reviewed (approved):
LGTM once we decide what we want to do for String values.
I see you already made the change to have them be extended JSON, so you'll either merge as is or simply revert back, in both case, no need to re-request review :)
Thank you for going through all of this. So many small details to take into account.

mattChiaravalloti commented:
evergreen merge

mattChiaravalloti commented:
evergreen merge

mattChiaravalloti commented:
just realized the commit queue task went purple the other day (system failure). resubmitted.

mattChiaravalloti commented:
@rychipman It looks like this cannot be merged until you approve since your last comment moved you to a "Requested Changes" state. (I attempted to remove you as a reviewer, but that moved you from the yellow circle to the red box). From evergreen:
> 405 1 review requesting changes and 2 approving reviews by reviewers with write access.

rychipman reviewed (approved):
LGTM!

mattChiaravalloti commented:
evergreen merge
